extensions/authentication-jwt/src/models/user.model.ts

```diff
@@ -54,6 +54,11 @@ export class User extends Entity {
   })
   verificationToken?: string;
 
+  @property.array(String, {
+    type: 'array',
+  })
+  roles: string[];
+
```

Comment:
How will this translate in relational databases? It would be better to create a `hasAndBelongsToMany` relation so that it follows SQL paradigms, and allows for easier maintenance and predictability.

Reply:
How will this work for authentication setups where there are no user roles and permissions are handled differently, e.g. via [OAuth2 scopes](https://oauth.net/2/scope/)? 

I think we need to find a way how to implement roles as an optional extension (a plugin).

Reply:
`Roles` and `Scopes` are not exclusive. The oAuth2 scopes represent the a list of permissions that a user grants to the client application. There are two levels of enforcement:

1. For a given request, does the request have enough user granted permissions (such as exposing email)?
2. Does the authenticated user have permission to access a protected resources? (a customer service to cancel an order for another user).

It would be nice to isolate the role based attributes into an extension, or at least, make it optional.

Reply:
> It would be better to create a hasAndBelongsToMany relation so that it follows SQL paradigms, and allows for easier maintenance and predictability.

We don't support dynamically define relation yet. Therefore it's hard to have a plugible role model. Even you can dynamically define a Role model +repository as extension, I am afraid you cannot define the relation.

Some sql connector like postgresql supports array field, e.g. [serialization of arrays](https://github.com/strongloop/loopback-connector-postgresql/pull/428/files)

How about add the `roles` property by mixn?
See tutorial [defining model property mixin](https://loopback.io/doc/en/lb4/migration-models-mixins.html#defining-the-model-property-mixin-class-factory-function)

This way you can also customize behaviour for sql connectors not supporting array field. Like storing roles as a string with different roles separate by `,` and authorizer parses it.

The authorizer can stay as is. 
The services can be extended. E.g. 

```ts
class UserWithRoleService extends MyUserService {
  constructor() {
    super();
  }
  // override to include roles
  convertToUserProfile(user: User): UserProfile {
    return {
      [securityId]: user.id.toString(),
      name: user.username,
      id: user.id,
      email: user.email,
      ...(user.roles && {roles: user.roles}),
    };
  }
}
```

The jwt.service.ts needs to extract user profile convert into a function, then you can override it easily.

Reply:
I agree with @raymondfeng regarding:

> It would be nice to isolate the role based attributes into an extension, or at least, make it optional.

So It would be much better to have another extension like lb4-acl @jannyHou @bajtos Can we create a scaffold of it, or a discussion on it to craft this feature.

Reply:
@madaky @odykyi That would be perfect. You can scaffold an extension using `lb4 extension`, see [instructions](https://loopback.io/doc/en/lb4/Extension-generator.html). The new extension can be based on `@loopback/authentication-jwt`, override/re-export classes from it.

let me know if you have questions.